wkcuber/api/Layer.py

```diff
@@ -99,20 +118,157 @@ def set_bounding_box_size(self, size: Tuple[int, int, int]) -> None:
         for _, mag in self.mags.items():
             mag.view.size = size
 
-    def setup_mag(self, mag: str) -> None:
+    def _initialize_mag_from_other_mag(
+        self, new_mag_name: Union[str, Mag], other_mag: MagDataset, compress: bool
+    ) -> MagDataset:
+        raise NotImplemented
+
+    def downsample(
+        self,
+        from_mag: Mag,
+        max_mag: Mag,
+        interpolation_mode: str,
+        compress: bool,
+        scale: Tuple[float, float, float] = None,
+        buffer_edge_len: int = None,
+        args: Namespace = None,
+    ) -> None:
+        # if 'scale' is set, the data gets downsampled anisotropic
+
+        # pad all existing mags if necessary
+        # during each downsampling step, the data shape or offset of the new mag should never need to be rounded
+        existing_mags = sorted([Mag(mag) for mag in self.mags.keys()])
+        all_mags_after_downsampling = existing_mags.copy()
+        cur_mag = get_next_mag(from_mag, scale)
+        while cur_mag <= max_mag:
+            all_mags_after_downsampling += [cur_mag]
+            cur_mag = get_next_mag(cur_mag, scale)
+
+        lowest_mag = all_mags_after_downsampling[-1].as_np()
+
+        # calculate aligned offset and size for lowest mag
+        offset_in_mag1 = self.dataset.properties.data_layers[
+            self.name
+        ].get_bounding_box_offset()
+        size_in_mag1 = self.dataset.properties.data_layers[
+            self.name
+        ].get_bounding_box_size()
+        offset_in_lowest_mag = offset_in_mag1 // all_mags_after_downsampling[-1].as_np()
+        end_offset_in_lowest_mag = -(
+            (np.array(offset_in_mag1) + size_in_mag1)
+            // -all_mags_after_downsampling[-1].as_np()
+        )  # ceil div
+        # translate alignment into mag 1
+        aligned_offset_in_mag1 = lowest_mag * offset_in_lowest_mag
+        aligned_size_in_mag1 = lowest_mag * (
+            end_offset_in_lowest_mag - offset_in_lowest_mag
+        )
+
+        # pad the existing mags
+        for mag_name in existing_mags:
```

Comment:
It would be cool if the padding-related code was extracted in a separate function and unit-tested. Do you think that's possible without too much effort?